ir/src/main/scala/org/scalajs/ir/Definitions.scala

```diff
@@ -89,190 +452,280 @@ object Definitions {
+            4
+          else
+            throwInvalidEncodedName(encoded)
+        }
+      }
     } else {
-      val pos = encodedName.indexOf("__")
-      (encodedName.substring(0, pos), encodedName.substring(pos + 2))
+      throwInvalidEncodedName(encoded)
     }
+  }
 
-    // -1 preserves trailing empty strings
-    val paramsAndResultStrings = privateAndSigString.split("__", -1).toSeq
-    val paramStrings :+ resultString = paramsAndResultStrings
+  // --- Encoding of strings into names ---
 
-    val paramTypes = paramStrings.map(decodeTypeRef).toList
-    val resultType =
-      if (resultString == "") None // constructor or reflective proxy
-      else Some(decodeTypeRef(resultString))
+  private def encodeStringName(name: String): EncodedName = {
+    // scalastyle:off return
+    val len = name.length()
+    if (len == 0)
+      throwEncodingError(name)
 
-    (simpleName, paramTypes, resultType)
-  }
-
-  /** Encodes a [[Types.TypeRef]] to be used for example in an encoded method
-   *  signature.
-   */
-  def encodeTypeRef(typeRef: TypeRef): String = {
-    typeRef match {
-      case PrimRef(tpe) =>
-        tpe match {
-          case NoType      => "V"
-          case BooleanType => "Z"
-          case CharType    => "C"
-          case ByteType    => "B"
-          case ShortType   => "S"
-          case IntType     => "I"
-          case LongType    => "J"
-          case FloatType   => "F"
-          case DoubleType  => "D"
-          case NullType    => "N"
-          case NothingType => "E"
-        }
-      case ClassRef(className) =>
-        className
-      case ArrayTypeRef(base, dimensions) =>
-        "A" * dimensions + encodeTypeRef(base)
+    /* We optimistically assume that all characters are ASCII, and backtrack if
+     * we find a non-ASCII character.
+     */
+    val result = new Array[Byte](len)
+    var i = 0
+    while (i != len) {
+      val c = name.charAt(i).toInt
+      if ((c & 0x7f) != c)
+        return encodeStringNameWithNonASCII(name)
+      result(i) = c.toByte
+      i += 1
     }
+    result
+    // scalastyle:on return
   }
 
-  /** Decodes a [[Types.TypeRef]], such as in an encoded method signature.
-   */
-  def decodeTypeRef(encodedName: String): TypeRef = {
-    val arrayDepth = encodedName.indexWhere(_ != 'A')
-    val base = {
-      if (arrayDepth == encodedName.length() - 1) {
-        (encodedName.charAt(arrayDepth): @switch) match {
-          case 'V' => VoidRef
-          case 'Z' => BooleanRef
-          case 'C' => CharRef
-          case 'B' => ByteRef
-          case 'S' => ShortRef
-          case 'I' => IntRef
-          case 'J' => LongRef
-          case 'F' => FloatRef
-          case 'D' => DoubleRef
-          case 'N' => NullRef
-          case 'E' => NothingRef
-          case _   => ClassRef(ClassName(encodedName.substring(arrayDepth)))
-        }
+  private def encodeStringNameWithNonASCII(name: String): EncodedName = {
```

Comment:
Just `String.getBytes(UTF_8)` is not correct here, because it does not report invalid inputs (unpaired surrogates). We have to use a `UTF_8.newEncoder()` manually and catch the `CharacterCodingException` that it throws, which is a bit of a pain. I've updated the code to use that, but I'm not sure the result is better. I can revert to the old version if you prefer.

Reply:
I like this much more, as it doesn't introduce encoding logic. Just boring code. I like boring code.